### PR TITLE
Convert NanGuardMode to use the VM linker instead of WrapLinker.

### PR DESCRIPTION
**Pull request by abergeron**
This removes the check on inputs that NanGuardMode used to do, but I feel it is not a big loss since you can check your function inputs easily.

### CODE REVIEW

File: theano/compile/nanguardmode.py [hunk 1]
```diff
@@ -277,36 +264,26 @@ def do_check_on(var, nd, f, is_input):
...
+            thunk : callable
                 The thunk to execute for this Apply node.
+            storage_map : dict
+                The mapping of variables to storage cells.
+            compute_map : dict
+                The mapping to get the computed state of the nodes. (ignored)
 
             """
-            inputs = fn.inputs
-            for x, var in zip(inputs, node.inputs):
```
Comment (nouiz) [inline theano/compile/nanguardmode.py]: Why did you remove the check on the input of the function?

Comment (abergeron) [inline theano/compile/nanguardmode.py]: Because the callback is called after the node has run and there might be some inplace stuff that would make it confusing.

Also, this only check the inputs of the graph so I figure it would be easier to check those yourself.

We can bring that back with a bit of modification to StackVM, but I didn't have the time to do that.

Comment (nouiz) [inline theano/compile/nanguardmode.py]: I don't like regression. So I would wait for this to be fixed before merging.

Comment (abergeron) [inline theano/compile/nanguardmode.py]: I just pushed a fix.

**Comment by nouiz** [comment]
travis fail.

File: theano/gof/vm.py [hunk 2]
```diff
@@ -378,6 +379,13 @@ def run_thunk_of_node(self, node):
 
         """
         idx = self.node_idx[node]
+        if self.callback_before is not None:
+            self.callback_before(
```
Comment (nouiz) [inline theano/gof/vm.py]: When we discussed this, I was more thinking that callback_before would be called before the function start execution, not before each node. This add extra overhead. I don't see a need for that case of callback before each node.

What do you think of that?

**Comment by abergeron** [comment]
I renamed callback_before to callback_input and made be called only on the inputs of the graph.

File: theano/compile/nanguardmode.py [hunk 3]
```diff
@@ -256,18 +251,14 @@ def do_check_on(var, nd, f, is_input):
...
                     error = True
             if error:
-                if not is_input:
-                    print("NanGuardMode found an error in the"
-                          " output of a node in this variable:", file=sio)
+                if nd:
+                    print("NanGuardMode found an error in the "
+                          "output of a node in this variable:", file=sio)
                     print(theano.printing.debugprint(nd, file='str'), file=sio)
+                    msg = sio.getvalue()
```
Comment (nouiz) [inline theano/compile/nanguardmode.py]: Why this line moved? I don't think it should have moved.

**Comment by nouiz** [comment]
travis fail.

**Comment by nouiz** [comment]
travis fail.

**Comment by abergeron** [comment]
This should be fixed.

File: theano/gof/vm.py [hunk 4]
```diff
@@ -406,6 +407,8 @@ def __call__(self, output_subset=None):
 
         for k in self.storage_map:
             compute_map[k][0] = (k.owner is None)
+            if self.callback_input and compute_map[k][0]:
+                self.callback_input(k, self.storage_map[k][0])
```
Comment (nouiz) [inline theano/gof/vm.py]: This behave as the old code. I just taught that this also check constant. I would be fine to only test the inputs of the fgraph and not constant. Otherwise, update the docstring around line 688 to tell this.

Comment (abergeron) [inline theano/gof/vm.py]: Done.

**Comment by lamblin** [comment]
We will need to update the doc as well, for instance http://deeplearning.net/software/theano/tutorial/modes.html?highlight=nanguardmode#linkers

**Comment by abergeron** [comment]
I did a git grep in the doc/ directory and this appears to be the only thing that changed.

**Comment by nouiz** [comment]
ok to merge when travis pass.
